ducktape/tests/action.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2015 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Provides pass/fail/aux-results reporting for specific test actions without
+# necessarily affecting the entire test outcome.
+#
+
+import traceback
+import inspect
+import time
+
+class TestAction (object):
+    # Accumulated test actions in various forms.
+    actions_root = list()
+    actions = list()
+    actions_by_fullname = dict()
+
+    def __init__ (self, name, instance=None, desc=None, parent=None, logger=None):
+        """
+            name     := Appended to parent's name to create unique action identifier
+            instance := Instance to test: allows multiple actions with the same name
+            desc     := Human readable description of this action
+            parent   := Parent action, either a TestAction object or a fully qualified
+                        action name (as returned by TestAction.fullname).
+            logger   := Logger to use. Only set this on your top-level action, it will be
+                        inherited to all sub-actions.
+        """
+        
+        super(TestAction, self).__init__()
+        self.name       = name
+        self.instance   = instance
+        self.desc       = desc
+        self.logger     = logger
+        self.start_time = time.time()
+        self.duration   = 999.999
+        # Resolve parent
+        if type(parent) == str:
```

Comment:
If you're going to check like this, `isinstance(parent, basestring)` is better here (and if we care about Python 3 you may need `six.string_types`, though so far we haven't worried about that).

But also, I'm also not sure about the logic here. If they specify a parent as a string and it doesn't exist in `actions`, is just removing the parent information what we want to do? Shouldn't that be considered an error? Also, is the lookup in `actions`, which seems to be a list of `TestActions` right? Or should we be looking it up in `actions_by_fullname`?